crates/services/p2p/src/discovery/discovery_config.rs

```diff
@@ -167,13 +162,42 @@ impl Config {
             MdnsWrapper::disabled()
         };
 
-        Behaviour {
+        Ok(Behaviour {
             connected_peers: HashSet::new(),
             kademlia,
             next_kad_random_walk,
             duration_to_next_kad: Duration::from_secs(1),
             max_peers_connected,
             mdns,
-        }
+        })
     }
 }
+
+async fn get_kademlia_kv(
+    multiaddrs: Vec<Multiaddr>,
+) -> anyhow::Result<Vec<(PeerId, Multiaddr)>> {
+    let dnsaddr_urls = multiaddrs
+        .iter()
+        .filter_map(|node| {
+            if let Protocol::Dnsaddr(multiaddr) = node.iter().next()? {
+                Some(multiaddr.clone())
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let dns_resolver = DnsResolver::new().await?;
+    let mut dnsaddr_multiaddrs = vec![];
+
+    for dnsaddr in dnsaddr_urls {
+        let multiaddrs = dns_resolver.lookup_dnsaddr(dnsaddr.as_ref()).await?;
```

Comment:
What if `dnsaddr` will have another `dnsaddr`, will it work?

Reply:
yup, resolution is recursive. the test case handles that :)